File: apps/native/mining-ui/macos-swiftui/KuberCoinMiner/Services/MiningService.swift
import Foundation

class MiningService {
    private let rpcUrl: String
    private let apiKey: String
    
    init() {
        self.rpcUrl = ProcessInfo.processInfo.environment["KUBERCOIN_RPC_URL"] ?? "http://127.0.0.1:8332"
        self.apiKey = ProcessInfo.processInfo.environment["KUBERCOIN_API_KEY"] ?? ""
    }
    
    func startMining() async throws {
        try await sendRpcCommand(method: "startmining", params: [])
    }
    
    func stopMining() async throws {
        try await sendRpcCommand(method: "stopmining", params: [])
    }
    
    private func sendRpcCommand(method: String, params: [Any]) async throws {
        guard !apiKey.isEmpty else {
            throw NSError(
                domain: "Missing API key",
                code: -2,
                userInfo: [NSLocalizedDescriptionKey: "Missing env var: KUBERCOIN_API_KEY"]
            )
        }

        guard let url = URL(string: rpcUrl) else {
            throw NSError(domain: "Invalid URL", code: -1)
        }
        
        let requestBody: [String: Any] = [
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1
        ]
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
        
        let (_, _) = try await URLSession.shared.data(for: request)
    }
}

File: apps/native/mining-ui/windows-winui3/Services/MiningService.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace KuberCoinMiner.Services
{
    public class MiningService
    {
        private readonly HttpClient _httpClient;
        private readonly string _rpcUrl;
        private bool _isMining;

        public MiningService()
        {
            _httpClient = new HttpClient();
            _rpcUrl = Environment.GetEnvironmentVariable("KUBERCOIN_RPC_URL")
                ?? throw new InvalidOperationException("Missing env var: KUBERCOIN_RPC_URL");

            var apiKey = Environment.GetEnvironmentVariable("KUBERCOIN_API_KEY")
                ?? throw new InvalidOperationException("Missing env var: KUBERCOIN_API_KEY");

            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {apiKey}");
            _httpClient.DefaultRequestHeaders.Add("X-API-Key", apiKey);
        }

        public async Task StartMiningAsync()
        {
            _isMining = true;
            // Start mining through RPC
            await SendRpcCommandAsync("startmining");
        }

        public async Task StopMiningAsync()
        {
            _isMining = false;
            // Stop mining through RPC
            await SendRpcCommandAsync("stopmining");
        }

        public bool IsMining => _isMining;

        private async Task<string> SendRpcCommandAsync(string method, params object[] parameters)
        {
            var request = new
            {
                jsonrpc = "2.0",
                method = method,
                @params = parameters,
                id = 1
            };

            var json = JsonSerializer.Serialize(request);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            try
            {
                var response = await _httpClient.PostAsync(_rpcUrl, content);
                return await response.Content.ReadAsStringAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"RPC Error: {ex.Message}");
                return string.Empty;
            }
        }
    }
}
